### DISCUSSION TITLE
鼠须管的布局属性让人感到困惑，需要更符合新版本的文档

### DISCUSSION THREAD

**Discussion by pdog18** [rime/home#1271] [Q&A] [answered]
鼠须管中的布局属性中有这几个属性
```
  corner_radius: 10
  hilited_corner_radius: 0
  border_height: 0
  border_width: 0
```
其中的 `border_height` 和 `border_width` 很容易让玩家认为，可以定义出一个垂直方方向与水平方向边框尺寸不相同的 border(例如 `top&bottom: 20px ,left&right : 10px`)。

但是实际在源码中的体现是:
```objective-c
// https://github.com/rime/squirrel/blob/7de3919aead2e425e66d3bacd3e5e441617cb66d/SquirrelPanel.m#L1988-L1993
  [theme setCornerRadius:cornerRadius
     hilitedCornerRadius:hilitedCornerRadius
       srdExtraExpansion:surroundingExtraExpansion
         shadowSize:shadowSize
               edgeInset:edgeInset
             borderWidth:MIN(borderHeight, borderWidth)
```

可以看到，border 总是选择其中较小的一个尺寸。
并且在实际测试中，如果两个尺寸不相等，还会将尺寸的插值计算至候选词对应方向的边距，这真的有些让人感到困惑了，不知道为何要这样设计?

**Comment by pdog18** [comment]
同时，鼠须管的 `corner_radius` 会影响到 `edgeInset` 属性，这是在 `macOS` 上的惯用风格吗?
例如一个 button 因为 `corner_radius` 的尺寸改变，直接会导致整个 `button` 的大小发生改变，有些让人感到意外了。

**Reply by LEOYoon-Tsaw** [reply]
不變的話圓角會切掉四角的字啊

**Reply by pdog18** [reply]
> 不變的話圓角會切掉四角的字啊

如果仅仅考虑这个原因就直接加上了 edgeInset 是否有些奇怪?
如果圆角会裁剪掉 4 角的字，额外使用 padding 之类的值是否更合适?
使用现在的方案 `corner_radius` 直接影响了 edgeInset 的值，那么用户想要生成一个圆角的同时，将文字内容更靠近边框，似乎需要不得不将 `border_width` 或 `border_height` 设置一个为负值。这就是让用户感到困惑的地方了。

**Reply by LEOYoon-Tsaw** [reply]
設置圓角相應增大邊距：一般情況下用戶不會遇到文字突出圓角之外的情況
負邊距可以強行縮小帶圓角時的邊距：用戶依然有辦法獨立控制邊距（but proceed at your own risk）

**Reply by pdog18** [reply]
感谢您的回复，理解这样设计的原因了。
(感觉有点反直觉，即使有 [Squirrel Designer](https://github.com/LEOYoon-Tsaw/Squirrel-Designer) 这么好的工具，我相信大多数人都无法在第一时间使用的很自然)

**Reply by LEOYoon-Tsaw** [reply]
其實這更多是一個歷史遺留問題
一開始Squirrel的圓角就是會增大邊距的（印象中從有圓角的時候就這樣），但是不允許設置負的邊距
我也喜歡窄邊，但是不能打破已有的格局，保持（基本的）向後兼容是鼠鬚管的一貫原則，於是我只能允許負邊距，來滿足想要圓角，同時又要窄邊的用戶

**Reply by pdog18** [reply]
理解「其實這更多是一個歷史遺留問題」，这也是无奈之举。
只是 `border_width` 与 `border_height` 同时承担了两项职责:
1. 控制 border 的尺寸
2. 控制文字与 border 的距离
对于需要「大圆角+窄边」的用户，应该还是需要摸索一下才能制作出「想要圓角，同時又要窄邊的」主题了。

另外说一个题外话:
同样表示圆角的 `hilited_corner_radius` 采用的就是不增加边距的规则，这才符合前端开发的惯例嘛(笑)。
另外有一点，同样关于`hilited_corner_radius`，鼠须管这里的 `hilited_corner_radius`如果是 `0` 的话，就会使得 `hilited_candidate_back_color` 之类的背景颜色侵入至 `corner_radius` 的区域，显得很酷。
例如这样:
![iShot_2023-04-22_12 16 01](https://user-images.githubusercontent.com/17060376/233761927-82a97484-0f07-48af-9fa3-15ad6001404a.gif)

**Reply by LEOYoon-Tsaw** [reply]
hilited_corner_radius是採取縮小自己大小的方式實現圓角，不過這些如果你覺得另一種邏輯更好，都可以改吧，我沒太大意見
至於最後那個，確實啊～

**Comment by LEOYoon-Tsaw** [comment, marked as answer]
我做了[Squirrel Designer](https://github.com/LEOYoon-Tsaw/Squirrel-Designer)，比任何文檔都強
